js/style/style_layer.js

```diff
@@ -114,15 +156,10 @@ StyleLayer.prototype = {
         }
     },
 
-    getPaintValue: function(name, zoom, zoomHistory) {
-        return this._transitions[name].at(zoom, zoomHistory);
-    },
-
     // update zoom
     recalculate: function(zoom, zoomHistory) {
-        this.paint = new PaintProperties[this.type]();
-
-        for (var name in this._transitions) {
+        this.paint = {};
+        for (var name in this._paintSpecifications) {
```

Comment:
`PaintProperties` doesn't make much sense alongside data-driven styles. If `circle-color` is a data-driven property, what does `paint['circle-color']` mean? I don't have a firm plan yet, but inlining the `PaintProperties` class will put us in the frame of mind to implement the right architecture when the time comes. 

Perhaps the FPS benchmark is a not a good test because it does not exercise zooming (which is what this would affect). However this is not especially hot code and we have `O(number of style spec properties)` operations in the same code path (the `draw_*` methods). 

I promise to do very thorough style benchmarking and profiling as part of https://github.com/mapbox/mapbox-gl-js/pull/1932.